T0103 invoice payment order restrictions.

simgonzalez opened the pull request:
Add restrictions for the payment on payment order.

ecino commented on recurring_contract/models/move.py:
```diff
@@ -139,6 +139,9 @@ def update_open_invoices(self, updt_val):
                                      and i.invoice_date_due >= date_selection):
+            # When the invoice has a payment order
+            if invoice.payment_order_id.filtered(lambda p: p.state not in ["draft", "open"]):
+                continue
```
This could simply be included in the filtered conditions in the loop above.

ecino commented on invoice_restrictions/models/account_move.py:
```diff
@@ -0,0 +1,22 @@
+    _inherit = 'account.move'
+
+    def write(self, vals):
+        if self.mapped('payment_order_id').filtered(lambda p: p.state not in ["draft", "open"]):
```
Should we filter for specific vals change only? Wouldn't some changes without impact on the payment order still be possible? I feel this global way is too preventive and might raise errors for cases that are not problematic in the end.

ecino commented on invoice_restrictions/__manifest__.py:
```diff
@@ -0,0 +1,42 @@
+#                        /_/
+#                            in Jesus' name
+#
+#    Copyright (C) 2014-2019 Compassion CH (http://www.compassion.ch)
+#    @author: Cyril Sester <csester@compassion.ch>, Emanuel Cino
```
It's not really reflecting the reality.

simgonzalez commented:
The rules must be rethought because it's not as easy as it looks and we have to define rules for exceptional case. Like taking out of the payment order the move and changing his payment_state from not paid to paid. and ...

simgonzalez commented on invoice_restrictions/models/account_move.py:
```diff
@@ -0,0 +1,40 @@
+    def write(self, vals):
+        move_to_modify = self.filtered('line_ids.payment_line_ids')
+        if move_to_modify:
+            payment_line_ids = move_to_modify.mapped('line_ids.payment_line_ids')
```
test the status in vals.

Commit:
T103 invoice payment restrictions.
Make the restrictions great again.
Use a behaviour that could be implemented in the OCA module and make sure it is well done.

ecino committed:
Merge branch '14.0' into T0103_restriction_for_invoices_in_DD
